### PR TITLE
feat(iotevents): allow setting description, evaluation method and key of DetectorModel

### PR DESCRIPTION
**Pull request by yamatatsu**
This PR is about https://github.com/aws/aws-cdk/issues/17711 (but out of the roadmap).
This PR (especially `key` property) make it easier to test the features we will implement.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-iotevents/README.md [hunk 1]
```diff
@@ -51,7 +51,7 @@ import * as iotevents from '@aws-cdk/aws-iotevents';
 
 const input = new iotevents.Input(this, 'MyInput', {
   inputName: 'my_input', // optional
-  attributeJsonPaths: ['payload.temperature'],
+  attributeJsonPaths: ['payload.deviceId', 'payload.temperature'],
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/README.md:54]: Why this change - seems unrelated to the rest of this PR...?

File: packages/@aws-cdk/aws-iotevents/lib/detector-model.ts [hunk 2]
```diff
@@ -16,6 +16,22 @@ export interface IDetectorModel extends IResource {
...
 
+/**
+ * Information about the order in which events are evaluated and how actions are executed.
+ */
+export enum EvaluationMethod {
+  /**
+   * When setting to SERIAL, variables are updated and event conditions are evaluated in the order
+   * that the events are defined.
+   */
+  BATCH = 'BATCH',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/lib/detector-model.ts:27]: Let's add an empty line after this line.

File: packages/@aws-cdk/aws-iotevents/lib/detector-model.ts [hunk 3]
```diff
@@ -16,6 +16,22 @@ export interface IDetectorModel extends IResource {
   readonly detectorModelName: string;
 }
 
+/**
+ * Information about the order in which events are evaluated and how actions are executed.
+ */
+export enum EvaluationMethod {
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/lib/detector-model.ts:22]: Is this name too generic? Should this be `EventEvaluationMethod`, or `StateEvaluationMethod`, or `DetectorActionEvaluationMethod`, or something?

Actually, can we even kill the "Method" suffix here? I don't think it adds much, and just makes the name longer.

File: packages/@aws-cdk/aws-iotevents/lib/detector-model.ts [hunk 4]
```diff
@@ -27,6 +43,38 @@ export interface DetectorModelProps {
    */
   readonly detectorModelName?: string;
 
+  /**
+   * A brief description of the detector model.
+   *
+   * @default none
+   */
+  readonly detectorModelDescription?: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/lib/detector-model.ts:51]: Let's call this `description` instead.

File: packages/@aws-cdk/aws-iotevents/lib/detector-model.ts [hunk 5]
```diff
@@ -27,6 +43,38 @@ export interface DetectorModelProps {
...
+   * detector instance with a unique key value is created. AWS IoT Events can continue to route
+   * input to its corresponding detector instance based on this identifying information.
+   *
+   * This parameter uses a JSON-path expression to select the attribute-value pair in the message
+   * payload that is used for identification. To route the message to the correct detector instance,
+   * the device must send a message payload that contains the same attribute-value.
+   *
+   * @default - none (single detector instance will be created and all inputs will be routed to it)
+   */
+  readonly key?: string;
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/lib/detector-model.ts:76]: This name, however, is way, way too generic. How about calling it `detectorKey`, or something like that?

File: packages/@aws-cdk/aws-iotevents/test/detector-model.test.ts [hunk 6]
```diff
@@ -95,6 +95,65 @@ test('can set physical name', () => {
...
+        condition: iotevents.Expression.fromString('test-eventCondition'),
+      }],
+    }),
+  });
+
+  // THEN
+  Template.fromStack(stack).hasResourceProperties('AWS::IoTEvents::DetectorModel', {
+    Key: 'payload.deviceId',
+  });
+});
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/test/detector-model.test.ts:155]: Let's just do all 3 in one test. Spreading them out like that just makes the tests longer, it doesn't really improve coverage in any way.

File: packages/@aws-cdk/aws-iotevents/test/integ.detector-model.ts [hunk 7]
```diff
@@ -7,7 +7,7 @@ class TestStack extends cdk.Stack {
 
     const input = new iotevents.Input(this, 'MyInput', {
       inputName: 'test_input',
-      attributeJsonPaths: ['payload.temperature'],
+      attributeJsonPaths: ['payload.deviceId', 'payload.temperature'],
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/test/integ.detector-model.ts:10]: Again, why are we changing existing things?

File: packages/@aws-cdk/aws-iotevents/test/integ.detector-model.ts [hunk 8]
```diff
@@ -27,6 +27,9 @@ class TestStack extends cdk.Stack {
 
     new iotevents.DetectorModel(this, 'MyDetectorModel', {
       detectorModelName: 'test-detector-model',
+      detectorModelDescription: 'test-detector-model-description',
+      evaluationMethod: iotevents.EvaluationMethod.SERIAL,
+      key: 'payload.deviceId',
```
Comment (skinny85) [inline packages/@aws-cdk/aws-iotevents/test/integ.detector-model.ts:32]: Can't we use `payload.temperature` here?

**Review by skinny85 — DISMISSED** [review]
Thanks @yamatatsu! Minor notes, mainly about naming.

File: packages/@aws-cdk/aws-iotevents/lib/detector-model.ts (hunk 3 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iotevents/lib/detector-model.ts:22]: `EventEvaluation` is good I think!

File: packages/@aws-cdk/aws-iotevents/lib/detector-model.ts (hunk 5 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iotevents/lib/detector-model.ts:76]: `detectorKey` is good! Hmm... I should think for more understandability instead of CFn copy. Thank you!

File: packages/@aws-cdk/aws-iotevents/README.md (hunk 1 above)
Comment (yamatatsu) [reply packages/@aws-cdk/aws-iotevents/README.md:54]: I modified this for the sake of reality.
The path passed in `detectorKey` must be declared in this `attributeJsonPaths`. And the value passed in `detectorKey` is often the unique value of the device or device group.
This is because IoT Events is often used to monitor devices and device groups.

I modified integ test for the same reason.

**Review by skinny85 — APPROVED** [review]
Looks great @yamatatsu - short and sweet, just the way I like it 😉.
